libs/datasets/dataset_orchestrator.py

```diff
@@ -0,0 +1,318 @@
+        )
+
+
+@dataclass
+class MultiRegionOrchestrator:
+    """Class to orchestrate dataset updates through state-focused parallelization."""
+
+    regions: Iterable[OneStateDataset]
+    cbsa_aggregator: statistical_areas.CountyToCBSAAggregator
+    region_overrides: Dict
+    print_stats: bool
+
+    @classmethod
+    def from_bulk_mrds(
+        cls,
+        states: Optional[List[str]] = None,
+        refresh_datasets: Optional[bool] = True,
+        print_stats: bool = True,
+    ) -> "MultiRegionOrchestrator":
+        bulk_dataset = load_bulk_dataset(refresh_datasets=refresh_datasets)
+        if not states:
+            states = list(US_STATE_ABBREV.values())
+        regions = [
+            OneStateDataset.from_mrds(bulk_dataset.get_subset(state=region)) for region in states
+        ]
+
+        cbsa_aggregator = statistical_areas.CountyToCBSAAggregator.from_local_public_data()
+        region_overrides = json.load(open(REGION_OVERRIDES_JSON))
+        return MultiRegionOrchestrator(
+            regions=regions,
+            region_overrides=region_overrides,
+            cbsa_aggregator=cbsa_aggregator,
+            print_stats=print_stats,
+        )
+
+    def build_and_combine_regions(
+        self, aggregate_to_country: bool = True, generate_cbsas: bool = True
+    ):
+        processed_regions = parallel_map(self._build_region_timeseries, self.regions)
+        _log.info("Finished processing individual regions...")
+        return self._combine_regions(
+            list(processed_regions),
+            aggregate_to_country=aggregate_to_country,
+            generate_cbsas=generate_cbsas,
+        )
+
+    def _combine_regions(
+        self,
+        datasets: List[MultiRegionDataset],
+        aggregate_to_country: bool = True,
+        generate_cbsas: bool = True,
+        generate_hsas: bool = True,
+    ) -> MultiRegionDataset:
+        regions = [region for dataset in datasets for region in dataset.location_ids]
+        assert len(regions) == len(set(regions)), "Can't combine datasets with duplicate locations"
+
+        _log.info("Starting region combination...")
+        timeseries_df = (
+            pd.concat([dataset.timeseries_bucketed for dataset in datasets])
+            .sort_index()
+            .rename_axis(columns=PdFields.VARIABLE)
+        )
+        static_df = (
+            pd.concat(dataset.static for dataset in datasets)
+            .sort_index()
+            .rename_axis(columns=PdFields.VARIABLE)
+        )
+        tag = pd.concat([dataset.tag for dataset in datasets]).sort_index()
+        multiregion_dataset = MultiRegionDataset(
+            timeseries_bucketed=timeseries_df, static=static_df, tag=tag
+        )
+
+        if generate_hsas:
+            _log.info("Starting HSA aggregation...")
+            hsa_aggregator = statistical_areas.CountyToHSAAggregator.from_local_data()
+            multiregion_dataset = hsa_aggregator.aggregate(multiregion_dataset)
+
+        if generate_cbsas:
```

Comment:
Out of curiosity do we ever not set generate_hsas and generate_cbsas to True?  Do we need to have these flags? 🤷

Reply:
They can be helpful for testing things quickly, as the cbsas/hsa aggregations make up a fair amount of the time of the whole pipeline, so while we don't ever use them when actually running the pipeline I'm biased towards leaving them in

Reply:
Ah, gotcha! 👍